hugr-passes/src/static_circuit.rs

```diff
@@ -0,0 +1,398 @@
+        }
+    }
+}
+
+impl<'a, H: HugrView> ConstLoader<QbHistory> for StaticCircuitContext<'a, H> {}
+
+impl<'a, H> Deref for StaticCircuitContext<'a, H> {
+    type Target = H;
+
+    fn deref(&self) -> &Self::Target {
+        self.hugr
+    }
+}
+
+impl<'a, H: HugrView> DFContext<QbHistory> for StaticCircuitContext<'a, H> {
+    type View = H;
+
+    fn interpret_leaf_op(
+        &self,
+        node: Node,
+        e: &ExtensionOp,
+        ins: &[PartialValue<QbHistory>],
+        outs: &mut [PartialValue<QbHistory>],
+    ) {
+        self.assert_invariants();
+        let extension = e.def().extension();
+        if !self.extensions.contains(extension) {
+            return;
+        }
+
+        if ins.iter().any(|x| x == &PartialValue::Bottom) {
+            return;
+        }
+
+        // assert_eq!(ins.len(), self.hugr.num_inputs(node));
+        // assert_eq!(outs.len(), self.hugr.num_outputs(node));
+
+        let qb_ins = qubits_in(self.hugr, node).collect::<Vec<_>>();
+        let qb_outs = qubits_out(self.hugr, node).collect::<Vec<_>>();
+
+        if qb_ins.is_empty() && qb_outs.is_empty() {
+            return;
+        }
+
+        if let Some(&qb_i) = self.alloc_ops.get(&node) {
+            assert_eq!(0, qb_ins.len());
+            assert_eq!(1, qb_outs.len());
+            for (i, out) in outs.iter_mut().enumerate() {
+                if qb_outs[0] == i.into() {
+                    out.join_mut(QbHistory::init(qb_i).into());
+                } else {
+                    out.join_mut(PartialValue::Top);
+                }
+            }
+        } else if self.free_ops.contains(&node) {
+            assert_eq!(0, qb_outs.len());
+            for out in outs.iter_mut() {
+                out.join_mut(PartialValue::Top);
+            }
+        } else {
+            assert_eq!(qb_outs.len(), qb_ins.len(), "{e:?}");
+            let mb_in_qbs = qb_ins
+                .iter()
+                .map(|&i| match &ins[i.index()] {
+                    PartialValue::Value(qbh) => Some((i, qbh)),
+                    _ => None,
+                })
+                .collect::<Option<HashMap<_, _>>>();
+            let gating_qbs = mb_in_qbs
+                .iter()
+                .flat_map(|hash_map| hash_map.iter().map(|(_, qbh)| qbh.qb_index()))
+                .collect_vec();
+            let qb_out_to_in: HashMap<_, _> = zip_eq(qb_outs, qb_ins).collect();
+            for (i, v) in outs.iter_mut().enumerate() {
+                let mb_join_val = (|| {
+                    let in_p = qb_out_to_in.get(&OutgoingPort::from(i))?;
+                    let hash_map = mb_in_qbs.as_ref()?;
+                    let qbh: &QbHistory = hash_map.get(in_p)?;
+                    let mut qbh = qbh.clone();
+                    qbh.push(node, e.name(), gating_qbs.clone());
+                    Some(qbh.into())
+                })();
+
+                v.join_mut(mb_join_val.unwrap_or(PartialValue::Top));
```

Comment:
Short comment - if  the corresponding in-port has a QbHistory, push this gate (+ ordered list of its inputs) onto that QbHistory...

But, if this in-port is classical, join with Top. Doesn't this hit `3.py` where you have two different classical values?

Reply:
How does one proceed in that case (given a TOP)? Is the goal here just to identify cases which could be statically scheduled if we then did a separate transformation pass?

Reply:
Yes, "then wire up the angles" is the "separate transformation" i.e. where you add extra edges, Output node ports, etc. to get the angles from wherever they were computed.